guidelines/creating_dockerfiles.md

```diff
@@ -0,0 +1,71 @@
+# Guidelines for creating Docker images for Kubeflow applications
+
+Authors: duselvar@cisco.com, jlewi@google.com, adselvaraj@cisco.com, jtf.github@gmail.com
+
+## Objective
+
+The purpose of this document is to outline a process for creating docker images maintained by our community for Kubeflow applications.
+
+The main goals are to: 
+
+1. Establish a process around creating images for Kubeflow applications.
+1. Improve security of the images built for running Kubeflow applications.
+
+## Scope
+
+This document is intended to cover container images of applications that are deployed in the
+Kubernetes cluster as part of Kubeflow and in particular:
+
+1. Custom resources and controllers
+1. Web applications
+
+This document is not intended to cover any image which is not deployed to the Kubernetes cluster.
+
+
+## Process
+
+Using [Multi-staged builds](https://docs.docker.com/develop/develop-images/multistage-build/) you can separate the build process to a separate build image and copy the binaries generated from the build stage to the actual image.
+
+### Selection of build image
+
+The recommended build images for each of the programming languages are listed below:
+
+| Programming Language or Framework  | Docker Image |
+| ------------- | ------------- |
+| Golang  | [Golang Docker Image](https://hub.docker.com/_/golang) |
+| Python3  | [Python Docker Image](https://hub.docker.com/_/python) |
+| NodeJS  | [NodeJS Docker Image](https://hub.docker.com/_/node/) |
+
+If your language or framework is not covered within these above recommendations, it is recommended to use the official images for the respective language or framework.
+
+#### Recommendation for Golang images
+
+- It is recommended to use Golang 1.13+ images while building docker images as it contains critical security fixes for the packages net/http, crypto/dsa and net/textproto.
+
+- While building Golang binaries in the build docker file, it is recommended to build with the below flags to ensure that the binary runs without needing to use `libc` based packages. This facilitates the use of distroless' `static` image as opposed to `base` image.
+```
+RUN CGO_ENABLED=0 GOOS=linux go build -o <output_binary> -ldflags "-w" -a .
```

Comment:
@krishnadurai should we be using `-w`?
What that flag does is that it strips debug symbols (DWARF tables).
https://golang.org/cmd/link/

This saves space so the final image will be smaller.
It will also limit the ability to debug the binary.
I would argue that it's better to have bigger images but with better debuggability in this case.
What do you think?

Reply:
@yanniszark I'm clarifying my understanding here.

This document is for building out production images. Would it help if the production images have DWARF tables? If yes, how so?

Reply:
@krishnadurai my understanding is that it would help with more advanced debugging options.
e.g, imagine your application has a bug and you want to see what is happening in the exact code path, step-by-step.
You can do that with delve, but without the DWARF tables, the function names will be meaningless symbols.
i don't have a strong opinion on this, either way is fine with me, but I thought I'd mention it.

Reply:
@yanniszark I see what you mean here. For debugging you would change the command (CMD) for the docker image which has the DWARF table.
It makes sense to retain it.